pkg/operator/boundsatokensignercontroller/controller.go

```diff
@@ -0,0 +1,296 @@
+	if err != nil && !errors.IsNotFound(err) {
+		return err
+	}
+	if errors.IsNotFound(err) {
+		certConfigMap = &corev1.ConfigMap{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: targetNamespace,
+				Name:      CertConfigMapName,
+			},
+			Data: map[string]string{},
+		}
+	}
+	currPublicKey := string(signingSecret.Data[PublicKeyKey])
+	hasThisKey := false
+	for _, value := range certConfigMap.Data {
+		if value == currPublicKey {
+			hasThisKey = true
+			break
+		}
+	}
+	if !hasThisKey {
+		// Increment until a unique name is found
+		nextKeyIndex := len(certConfigMap.Data) + 1
+		nextKeyKey := ""
+		for len(nextKeyKey) == 0 {
+			possibleKey := fmt.Sprintf("service-account-%03d.pub", nextKeyIndex)
+			_, ok := certConfigMap.Data[possibleKey]
+			if !ok {
+				nextKeyKey = possibleKey
+				break
+			}
+			nextKeyIndex += 1
+		}
+
+		certConfigMap.Data[nextKeyKey] = currPublicKey
+		certConfigMap, _, err = resourceapply.ApplyConfigMap(c.configMapClient, c.eventRecorder, certConfigMap)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Check if next-sa-signing-key has been around long enough to be promoted.
+	// Giving time for apiserver instances to pick up the change in public keys before
+	// changing the private key minimizes the potential for one or more apiservers to
+	// issue tokens signed by the new private key that apiservers without the
+	// corresponding public key are unable to validate.
+	//
+	// TODO(marun) Find a more accurate indication that all apiservers are capable of
+	// validating tokens signed by the new private key.
+	readyToPromote := false
+	readyTime := signingSecret.Annotations[TokenReadyAnnotation]
+	if len(readyTime) == 0 {
+		readyToPromote = true
+	} else {
+		promotionTime, err := time.Parse(time.RFC3339, readyTime)
+		if err != nil || time.Now().After(promotionTime) {
+			readyToPromote = true
+		}
+	}
+
+	// If past promotion time, make the new private key available to the operand.
+	if readyToPromote {
+		_, _, err := resourceapply.SyncSecret(c.secretClient, c.eventRecorder,
+			operatorNamespace, NextSigningKeySecretName,
+			targetNamespace, SigningKeySecretName, []metav1.OwnerReference{})
+		return err
+	}
+
+	return nil
+}
+
+func (c *BoundSATokenSignerController) Run(ctx context.Context) {
+	defer utilruntime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	klog.Infof("Starting BoundSATokenSignerController")
+	defer klog.Infof("Shutting down BoundSATokenSignerController")
+
+	if !cache.WaitForCacheSync(ctx.Done(), c.cachesSynced...) {
+		utilruntime.HandleError(fmt.Errorf("caches did not sync"))
+		return
+	}
+
+	stopCh := ctx.Done()
+
+	// Run only a single worker
+	go wait.Until(c.runWorker, time.Second, stopCh)
+
+	// start a time based thread to ensure we stay up to date
+	go wait.Until(func() {
+		ticker := time.NewTicker(time.Minute)
+		defer ticker.Stop()
+
+		for {
+			c.queue.Add(workQueueKey)
+			select {
+			case <-ticker.C:
+			case <-stopCh:
+				return
+			}
+		}
+
+	}, time.Minute, stopCh)
```

Comment:
It this correct? That inner `ticker` looks quite unnecessary given that `wait.Until` has its own internal timer which does the same, this basically runs `wait.Until` and freezes it on its first loop?

Reply:
I think you're right - it should be sufficient to call `func() {c.queue.Add(workQueueKey)}` as the argument to `wait.Util`. Updated.

Reply:
Done